gcsio/src/main/java/com/google/cloud/hadoop/gcsio/BatchExecutor.java

```diff
@@ -0,0 +1,144 @@
+ */
+class BatchExecutor {
+
+  private static final GoogleLogger logger = GoogleLogger.forEnclosingClass();
+
+  private final ExecutorService requestsExecutor;
+
+  private final Queue<Future<Void>> responseFutures = new ConcurrentLinkedQueue<>();
+
+  public BatchExecutor(int numThreads) {
+    this.requestsExecutor =
+        numThreads == 0 ? newDirectExecutorService() : newRequestExecutor(numThreads);
+  }
+
+  private static ExecutorService newRequestExecutor(int numThreads) {
+    ThreadPoolExecutor executor =
+        new ThreadPoolExecutor(
+            /* corePoolSize= */ numThreads,
+            /* maximumPoolSize= */ numThreads,
+            /* keepAliveTime= */ 10L,
+            TimeUnit.SECONDS,
+            new LinkedBlockingQueue<>(numThreads * 20),
+            new ThreadFactoryBuilder()
+                .setNameFormat("gcs-manual-batching-pool-%d")
+                .setDaemon(true)
+                .build());
+    executor.allowCoreThreadTimeOut(true);
+    executor.setRejectedExecutionHandler(new ThreadPoolExecutor.CallerRunsPolicy());
+    return executor;
+  }
+
+  /** Adds a task to the execution queue. */
+  public <T> void queue(Callable<T> task, FutureCallback<T> callback) {
+    checkState(
+        !requestsExecutor.isShutdown() && !requestsExecutor.isTerminated(),
+        "requestExecutor should not be terminated to queue request");
+
+    responseFutures.add(
+        requestsExecutor.submit(
+            () -> {
+              execute(task, callback);
+              return null;
+            }));
+  }
+
+  private static <T> void execute(Callable<T> task, FutureCallback<T> callback) {
+    checkArgument(callback != null, "FutureCallBack cannot be null : %s", callback);
+    try {
+      T result = task.call();
+      callback.onSuccess(result);
+    } catch (Throwable throwable) {
+      callback.onFailure(throwable);
+    }
+  }
+
+  /** Awaits until all tasks are terminated and then shutdowns the executor. */
+  public void shutdown() throws IOException {
+    try {
+      awaitRequestsCompletion();
+      checkState(responseFutures.isEmpty(), "responseFutures should be empty after flush");
+    } finally {
+      requestsExecutor.shutdown();
+      try {
+        if (!requestsExecutor.awaitTermination(1, TimeUnit.SECONDS)) {
+          logger.atWarning().log("Forcibly shutting down manual batching thread pool.");
+          requestsExecutor.shutdownNow();
+        }
+      } catch (InterruptedException e) {
+        Thread.currentThread().interrupt();
+        logger.atFine().withCause(e).log(
+            "Failed to await termination: forcibly shutting down manual batching thread pool.");
+        requestsExecutor.shutdownNow();
+      }
+    }
+  }
+
+  /** Awaits until all sent requests are completed */
+  private void awaitRequestsCompletion() throws IOException {
+    while (!responseFutures.isEmpty()) {
+      getFromFuture(responseFutures.remove());
```

Comment:
If there are multiple failures, we will get only the first exception, correct?

Reply:
the failure will throw exception but the rest of the processes will continue to be executed.

Reply:
Those exceptions will be lost, correct?